IPython/terminal/interactiveshell.py

```diff
@@ -444,9 +450,9 @@ def pre_prompt(self):
             # after the buffer is reset.
             s = self.rl_next_input
             def set_doc():
-                self.pt_cli.application.buffer.document = Document(s)
-            if hasattr(self.pt_cli, 'pre_run_callables'):
-                self.pt_cli.pre_run_callables.append(set_doc)
+                self.pt_app.application.buffer.document = Document(s)
+            if hasattr(self.pt_app.application, 'pre_run_callables'):
```

Comment:
IIRC this `hasattr()` check was for compatibility with older prompt_toolkit versions. As this will require 2.x anyway, we can probably remove it and only use `pre_run_callables`.

Reply:
Right. Fixed now.

Reply:
What is actually the use for this ``pre_prompt`` function? What is ``rl_next_input``?

Reply:
Various functions can set some text to appear at the next prompt, e.g. I can run `%load foo.py` to get the contents of `foo.py` at the next prompt. There's also `%recall` which can retrieve entries from history. That gets stored in `rl_next_input` and then inserted when the next prompt is being prepared.